packages/clay-shared/src/useFocusManagement.ts

```diff
@@ -0,0 +1,224 @@
+	'[tabindex]:not([tabindex^="-"])',
+	'area[href]',
+	'button:not([disabled]):not([aria-hidden])',
+	'embed',
+	'iframe',
+	'input:not([disabled]):not([type="hidden"]):not([aria-hidden])',
+	'object',
+	'select:not([disabled]):not([aria-hidden])',
+	'textarea:not([disabled]):not([aria-hidden])',
+];
+
+export function useFocusManagement(loop: boolean = false) {
+	const focusManagerRef = useRef<Array<FocusManager>>([]);
+	const lastPositionRef = useRef<number | null>(null);
+	const nextElementOutsideScopeRef = useRef<HTMLElement | null>(null);
+
+	const createScope = (ref: any, key: string, portal: boolean = false) => {
+		const elementIndex = focusManagerRef.current.findIndex(
+			el => (el && el.key) === key
+		);
+		const schema = {
+			key,
+			portal,
+			value: ref,
+		};
+
+		if (ref) {
+			if (elementIndex === -1) {
+				focusManagerRef.current.push(schema);
+			} else {
+				focusManagerRef.current[elementIndex] = schema;
+			}
+		} else {
+			if (lastPositionRef.current !== null) {
+				lastPositionRef.current = 0;
+			}
+			focusManagerRef.current.splice(elementIndex, 1);
+		}
+	};
+
+	const getFocusableElementOutsideScope = (
+		element: HTMLElement,
+		backwards: boolean
+	) => {
+		const firstElement = element;
+		element = element.parentNode as HTMLElement;
+
+		const match = () => {
+			if (!element || element.nodeType !== 1) {
+				return false;
+			}
+
+			const focusableElements = Array.prototype.slice.call(
+				element.querySelectorAll(FOCUSABLE_ELEMENTS as any)
+			);
+			const activeElementFocus = focusableElements.findIndex(
+				el => el === firstElement
+			);
+			const item = backwards
+				? activeElementFocus - 1
+				: activeElementFocus + 1;
+
+			if (focusableElements[item]) {
+				element = focusableElements[item];
+				return true;
+			}
+
+			return false;
+		};
+
+		while (element && !match()) {
+			element = element.parentNode as HTMLElement;
+		}
+
+		return element;
+	};
+
+	const getElementOutsidePortal = (element: FocusManager) => {
+		while (element && !element.portal === false) {
+			const index = focusManagerRef.current.findIndex(
+				el => el.key === element.key
+			);
+			element = focusManagerRef.current[index - 1];
+		}
+
+		return element;
+	};
+
+	const handleNextElementOutsideScope = (event: KeyboardEvent) => {
+		if (event.keyCode === TAB_KEY_CODE && event.shiftKey) {
+			event.preventDefault();
+			lastPositionRef.current = focusManagerRef.current.length;
+			moveFocusInScope(true);
+		}
+	};
+
+	const moveFocusInScope = (backwards: boolean = false) => {
+		const size = focusManagerRef.current.length - 1;
+		const activeNode = document.activeElement;
+		const firstPosition = lastPositionRef.current === 0;
+		const endPosition =
+			lastPositionRef.current !== null &&
+			size === lastPositionRef.current;
+
+		// Get the parent elements or focusable children of the
+		// non-portal element to move focus.
+		if (
+			!loop &&
+			(((lastPositionRef.current === null || firstPosition) &&
```

Comment:
This condition is pretty tough to follow, can we break it up into more readable chunks?

Reply:
Hmm yeah, I think I can simplify it more, I'll think about it.